### ISSUE TITLE
Feature Request: Add option to cleanup local files regardless of sync state when using --download-only

### ISSUE THREAD

**Issue by abraunegg** [abraunegg/onedrive#2092]
### Is your feature request related to a problem? Please describe.

When using `--download-only` files may be downloaded, but, depending on the sync state, will remain local even though they have been deleted online, thus, these need to be deleted locally rather than kept as a local archive.

### Describe the solution you'd like

Add option to cleanup local files regardless of sync state when using `--download-only`

### Describe alternatives you've considered

None

### Additional context

As per discussion here: https://github.com/abraunegg/onedrive/discussions/2087#discussioncomment-3387136

**Comment by rrodrigueznt** [comment]
Thanks for the post. The same applies to manually or programmatically locally created files: it would end up in a mess if they can not be clean-up to keep coherence with server contents.

**Comment by rrodrigueznt** [comment]
I'm happily concentrating in the _EXTRA_ warning of your message! First trials on a test machine doing great! I'm using a personal `OneDrive` account from an Ubuntu box:

```
rjr@IGFAE-MU-07-MacPro-Ubuntu:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.1 LTS
Release:        22.04
Codename:       jammy
```
What I don't find is logging information about deleted files. Please, read the following block:
```
rjr@IGFAE-MU-07-MacPro-Ubuntu:~$ ~/onedrive/onedrive --confdir="~/.config/onedrive_personal" --monitor --verbose --download-only --cleanup-local-files
Using 'user' Config Dir: /home/rjr/.config/onedrive_personal
Configuration file successfully loaded
Using logfile dir: /var/log/onedrive_personal/
Checking Application Version ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Using Curl defaults for all HTTP operations
Opening the item database ...
All operations will be performed in: /mnt/data01/OneDrive_personal
Application version: v2.4.20-31-g83ec0c1
Account Type: personal
Default Drive ID: a3e4920205347130
Default Root ID: A3E4920205347130!101
Remaining Free Space: 1697868469
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
WARNING: Application has been configured to cleanup local files that are not present online.
WARNING: Local data loss MAY occur in this scenario if you are expecting data to remain archived locally.
ERROR: Unsupported account type for syncing OneDrive Business Shared Folders
Initializing monitor ...
OneDrive monitor interval (seconds): 300
Starting a sync with OneDrive
Applying changes of Path ID: A3E4920205347130!101
Updated Remaining Free Space: 1697868469
Searching local filesystem for extra files and folders which need to be removed
Removing local directory as --download-only & --cleanup-local-files configured
Removing local directory: ./dummy_folder
Sync with OneDrive is complete
```
A first issue arising from these trials is that I don't get a log record when a file is deleted within a folder. You can see in the block below a record when the client removes a locally created copy of a `pdf` file, and another one when removing a locally created dummy folder. But no record reflecting the deletion of a second copy of the `pdf` file created within the dummy folder.
```
2022-Aug-29 12:13:14.8119278	Searching local filesystem for extra files and folders which need to be removed
2022-Aug-29 12:13:16.3701145	Removing local file as --download-only & --cleanup-local-files configured
2022-Aug-29 12:13:16.3702084	Removing local file: ./0000.Getting started with OneDrive (copy).pdf
2022-Aug-29 12:13:16.5473963	Removing local directory as --download-only & --cleanup-local-files configured
2022-Aug-29 12:13:16.5475053	Removing local directory: ./dummy_folder
2022-Aug-29 12:13:16.5480711	Sync with OneDrive is complete
```
Please, does it work that way by design? At least for me, it would be useful to keep track of any deleted document.

I'll do more tests on Ubuntu and CentOS Stream 8 and keep this thread updated! Thanks for your work!

**Comment by abraunegg** [comment]
@rrodrigueznt 
> Please, does it work that way by design? At least for me, it would be useful to keep track of any deleted document.

I made the design decision to only print out individual files if they are being removed.

For folders, if the parent is being removed, this is also printed out, but all 'children' of that folder being deleted, notification is being suppressed - as, I felt it was 100% redundant as the parent is what is being removed (thus, all children will be deleted) - better to notify that the parent object is what was removed, than everything under it + the parent ... 

From my perspective, the PR is operating as per design and expectation.

**Comment by rrodrigueznt** [comment]
@abraunegg 

> For folders, if the parent is being removed, this is also printed out, but all 'children' of that folder being deleted, notification is being suppressed - as, I felt it was 100% redundant as the parent is what is being removed (thus, all children will be deleted) - better to notify that the parent object is what was removed, than everything under it + the parent ...

I agree about this being the more economical solution regarding registered data, but I would prefer to store evidence of any deletion. Otherwise, it won't be possible to reconstruct the history of a given sub-folder or file being locally created by error or by intent.

Could you consider this behavior at least as an option? Thanks!

**Comment by abraunegg** [comment]
@rrodrigueznt 
> Could you consider this behavior at least as an option? Thanks!

Absolutely - as this is the whole point of a PR, to develop the required solution.

**Comment by abraunegg** [comment]
@rrodrigueznt 
Please can you rebuild your PR - you should have `onedrive v2.4.20-32-gf47974f` which will now output all files / folders that are being removed:
```
Searching local filesystem for extra files and folders which need to be removed
Removing local directory as --download-only & --cleanup-local-files configured
Removing local file: ./extra_folder/extra.file
Removing local file: ./extra_folder/folder_1/extra.file2
Removing local directory: ./extra_folder/folder_1
Removing local directory: ./extra_folder
```

**Comment by rrodrigueznt** [comment]
@abraunegg 
I was trying to understand how to update my local copy of the PR to test `onedrive v2.4.20-32-gf47974f` when I saw you pushed 1 commit fixing issue 2092. Sorry for being late, but I'm new to collaborative software development and I've to be very cautious with my actions!
Please, must I keep trying to rebuild my PR or update and rebuild master? I'm a bit lost with this! Thanks!

**Comment by abraunegg** [comment]
@rrodrigueznt 
To test the PR, you need to rebuild the PR as per the steps provided.

Once that PR is merged to 'master' you need to rebuild and install from 'master' to get the latest code. 

If you are using a package, wait for the next release and package to be updated.

 If you are using packages, you will need to perform a 'make uninstall' step to ensure any PR or 'master' version is removed from your local system.

**Comment by rrodrigueznt** [comment]
@abraunegg 
> @rrodrigueznt To test the PR, you need to rebuild the PR as per the steps provided.
> 
I think I understand now this message:
```
rjr@IGFAE-MU-07-MacPro-Ubuntu:~/onedrive$ gh pr checkout pr2113
no pull requests found for branch "pr2113"
rjr@IGFAE-MU-07-MacPro-Ubuntu:~/onedrive$
```
Does it mean that the PR was already merged to 'master'? Thanks!

**Comment by abraunegg** [comment]
@rrodrigueznt 
No - it means that you have already checked out the branch.

Suggest you delete that `./onedrive` folder and start again

**Comment by rrodrigueznt** [comment]
@abraunegg 
> Suggest you delete that `./onedrive` folder and start again
OK. I was trying to update my local copy of the PR.

**Comment by rrodrigueznt** [comment]
Done! It is working fine. Thanks!
```
2022-Aug-30 13:20:55.4740364	Starting a sync with OneDrive
2022-Aug-30 13:20:55.4742037	Applying changes of Path ID: A3E4920205347130!101
2022-Aug-30 13:20:56.5292327	Updated Remaining Free Space: 1692913812
2022-Aug-30 13:20:57.1773536	Searching local filesystem for extra files and folders which need to be removed
2022-Aug-30 13:20:58.790739	Removing local file as --download-only & --cleanup-local-files configured
2022-Aug-30 13:20:58.7908317	Removing local file: ./0000.Getting started with OneDrive (copy).pdf
2022-Aug-30 13:20:58.9737737	Removing local directory as --download-only & --cleanup-local-files configured
2022-Aug-30 13:20:58.9738836	Removing local file: ./dummy/0000.Getting started with OneDrive.pdf
2022-Aug-30 13:20:58.9740806	Removing local directory: ./dummy
2022-Aug-30 13:20:58.9744594	Sync with OneDrive is complete
```
For the record:
```
rjr@IGFAE-MU-07-MacPro-Ubuntu:~/onedrive$ ./onedrive --version
onedrive v2.4.20-34-g83ed2a6
```

**Comment by abraunegg** [comment]
@rrodrigueznt 
Thanks for the confirmation, will merge this into master.

**Comment by abraunegg** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
